### DISCUSSION TITLE
Using sis_user_id with enroll_user()

### DISCUSSION THREAD

**Discussion by DonDeWitt** [ucfopen/canvasapi#660] [Q&A]
The enroll_user() function of the Course class allows the enrollment of a user using a User object or the user's canvas id. From the docs:

enroll_user(user, enrollment_type=None, **kwargs)
Create a new user enrollment for a course or a section.

Calls:	
POST /api/v1/courses/:course_id/enrollments

Parameters:	
user (canvasapi.user.User or int) – The object or ID of the user to enroll in this course.
enrollment_type (str, optional) – The type of enrollment.
Return type:	
canvasapi.enrollment.Enrollment

To my understanding, the canvas API allows users to be referenced by their sis_user_id simply by [prefixing](https://community.canvaslms.com/t5/Canvas-Developers-Group/Enrollment-API-Questions/m-p/475337/highlight/true#M7389) the id with "sis_user_id:". Using the sis_user_id in my current situation would be a lot better, because I would not need to call the api for the user's object and canvas ID. However, from reading the code and documentation, it seems this is not possible to do with this canvasAPI? I am iterating over many students to enroll them in courses, and doing the call to get each one's data individually takes some time. I believe using the sis_user_id instead would speed things up for me, since I don't have the canvas user id's stored in my database.
Thanks.

**Comment by bennettscience** [comment]
The `enroll_user` function doesn't accept the `sis_user_id` option right now. You could build your own request using the `Requester` instance on the `canvas` object. It is attached to any object that can make a request in the library, so you can use it like this:

```python
from canvasapi.util import combine_kwargs
# assuming you have a CSV or something

def enroll_user(course_id, **kwargs)
    for user_id in sis_id_list:
        kwargs["enrollment[sis_user_id:user_id]"] = user_id
    
        # other params for the enrollment...
    
        response = canvas._requester.request(
            "POST",
            "courses/12345/enrollments",
            _kwargs=combine_kwargs(**kwargs),
        )
        return response.json()
```
Note that I've not done this myself, but it should get you close. The `Requester` object can be used for any call instead of the existing methods if they don't do what you need them to do.

**Reply by jonespm** [reply]
To run this outside of a class in the API I had to modify it to be `canvas._Canvas__requester`. This snippet worked as a GET request to retrieve the deleted users by sis_user_id. `canvas` was defined globally. 

```
def get_deleted_enrollments(user_id) -> PaginatedList:
    params={
        'state[]': 'deleted',
    }

    try:
        enrollments = PaginatedList(
            Enrollment,
            canvas._Canvas__requester,
            "GET",
            f"users/sis_user_id:{user_id}/enrollments",
            _kwargs=combine_kwargs(**params),
        )
    except CanvasException as e:
        logger.error("Error fetching enrollments for user {user_id}: {e}")
        return PaginatedList
    return enrollments
```

**Reply by ndegroot** [reply]
There is an open and updated pull request #440  which tries to fix this, waiting for review...
